tests/functional/olp-cpp-sdk-dataservice-read/VersionedLayerClientPrefetchPartitionsTest.cpp

```diff
@@ -0,0 +1,228 @@
+        mock_server_client_->MockGetResponse(
+            kLayer,
+            mockserver::ReadDefaultResponses::GenerateDataHandle(partitions[i]),
+            data);
+      }
+    }
+    std::promise<dr::PrefetchPartitionsResponse> promise;
+    auto future = promise.get_future();
+    auto token = client.PrefetchPartitions(
+        request, [&promise](dr::PrefetchPartitionsResponse response) {
+          promise.set_value(std::move(response));
+        });
+    ASSERT_NE(future.wait_for(kWaitTimeout), std::future_status::timeout);
+
+    auto response = future.get();
+    ASSERT_TRUE(response.IsSuccessful())
+        << response.GetError().GetMessage().c_str();
+    const auto result = response.MoveResult();
+
+    EXPECT_EQ(result.GetPartitions().size(), 10u);
+    for (const auto& partition : partitions) {
+      ASSERT_TRUE(client.IsCached(partition));
+
+      auto data_response =
+          client
+              .GetData(
+                  dr::DataRequest().WithPartitionId(partition).WithFetchOption(
+                      dr::FetchOptions::CacheOnly))
+              .GetFuture()
+              .get();
+      ASSERT_TRUE(data_response.IsSuccessful());
+      EXPECT_EQ(data_response.GetResult()->size(), 64u);
+    }
+    EXPECT_TRUE(mock_server_client_->Verify());
+  }
+  {
+    SCOPED_TRACE("Call get data than prefetch");
+    auto partition = "12";
+    const auto data = mockserver::ReadDefaultResponses::GenerateData();
+    const auto request =
+        dr::PrefetchPartitionsRequest().WithPartitionIds({partition});
+    {
+      mock_server_client_->MockGetResponse(
+          mockserver::ReadDefaultResponses::GeneratePartitionsResponse(1, 12),
+          GetPartitions());
+      mock_server_client_->MockGetResponse(
+          kLayer,
+          mockserver::ReadDefaultResponses::GenerateDataHandle(partition),
+          data);
+    }
+    auto data_future =
+        client.GetData(dr::DataRequest().WithPartitionId(partition))
+            .GetFuture();
+    ASSERT_NE(data_future.wait_for(kWaitTimeout), std::future_status::timeout);
+
+    auto data_response = data_future.get();
+    ASSERT_TRUE(data_response.IsSuccessful());
+    EXPECT_EQ(data_response.GetResult()->size(), data.size());
+
+    auto future = client.PrefetchPartitions(request).GetFuture();
+    ASSERT_NE(future.wait_for(kWaitTimeout), std::future_status::timeout);
+
+    auto response = future.get();
+    ASSERT_TRUE(response.IsSuccessful())
+        << response.GetError().GetMessage().c_str();
+    const auto result = response.MoveResult();
+
+    EXPECT_EQ(result.GetPartitions().size(), 1u);
+    ASSERT_TRUE(client.IsCached(partition));
+
+    EXPECT_TRUE(mock_server_client_->Verify());
+  }
+}
+
+TEST_F(VersionedLayerClientPrefetchPartitionsTest, PrefetchPartitionFail) {
+  olp::client::HRN hrn(kTestHrn);
+
+  auto client = dr::VersionedLayerClient(hrn, kLayer, boost::none, *settings_);
+  {
+    SCOPED_TRACE("Empty request");
```

Comment:
This is not necessary as it is the same test already done in integration test. Functional tests is the "e2e" test that covers the entire pipeline from user API to network layer. So test only those that reach the network and require an answer from the server.

E.g. 200 partitions, 1 query fails the other succeeds.